// @discordx/lava-player/build/esm/base/Node.js
/* eslint-disable @typescript-eslint/no-explicit-any */
import { EventEmitter } from "events";
import WebSocket from "ws";
import { Connection } from "../core/Connection.js";
import { Http } from "../core/Http.js";
import PlayerStore from "../core/PlayerStore.js";
export class BaseNode extends EventEmitter {
    password;
    userId;
    shardCount;
    connection;
    players = new PlayerStore(this);
    http;
    voiceStates = new Map();
    voiceServers = new Map();
    _expectingConnection = new Set();
    constructor({ password, userId, shardCount, host }) {
        super();
        this.password = password;
        this.userId = userId;
        this.shardCount = shardCount;
        const restIsSecure = host?.rest?.secure ?? host?.secure ?? false;
        const restAddress = host?.rest?.address ?? host?.address ?? "localhost";
        const restPort = host?.rest?.port ?? host?.port ?? 2333;
        this.http = new Http(this, `${restIsSecure ? "https" : "http"}://${restAddress}:${restPort}`);
        const wsIsSecure = host?.secure ?? false;
        const wsAddress = host?.address ?? "localhost";
        const wsPort = host?.port ?? 2333;
        this.connection = new Connection(this, `${wsIsSecure ? "wss" : "ws"}://${wsAddress}:${wsPort}`, host?.connectionOptions);
    }
    get connected() {
        return this.connection?.ws.readyState === WebSocket.OPEN;
    }
    load(identifier) {
        if (this.http) {
            return this.http.load(identifier);
        }
        throw new Error("no available http module");
    }
    decode(tracks) {
        if (this.http) {
            return this.http.decode(tracks);
        }
        throw new Error("no available http module");
    }
    voiceStateUpdate(packet) {
        if (packet.user_id !== this.userId) {
            return Promise.resolve(false);
        }
        if (packet.channel_id) {
            this.voiceStates.set(packet.guild_id, packet);
            return this._tryConnection(packet.guild_id);
        }
        else {
            this.voiceServers.delete(packet.guild_id);
            this.voiceStates.delete(packet.guild_id);
        }
        return Promise.resolve(false);
    }
    voiceServerUpdate(packet) {
        this.voiceServers.set(packet.guild_id, packet);
        this._expectingConnection.add(packet.guild_id);
        return this._tryConnection(packet.guild_id);
    }
    connect() {
        this.connection?.connect();
    }
    disconnect(code, data) {
        if (this.connection) {
            return this.connection.close(code, data);
        }
        return Promise.resolve();
    }
    async destroy(code, data) {
        await Promise.all([...this.players.values()].map((player) => player.destroy()));
        await this.disconnect(code, data);
    }
    async _tryConnection(guildId) {
        const state = this.voiceStates.get(guildId);
        const server = this.voiceServers.get(guildId);
        if (!state || !server || !this._expectingConnection.has(guildId)) {
            return false;
        }
        await this.players.get(guildId).voiceUpdate(state.session_id, server);
        this._expectingConnection.delete(guildId);
        return true;
    }
}
